maven-extension/src/main/java/io/opentelemetry/maven/OtelExecutionListener.java

```diff
@@ -0,0 +1,248 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Close the OpenTelemetry SDK (see {@link OpenTelemetrySdkService#dispose()} on the end of
+ * execution of the last project ({@link #projectSucceeded(ExecutionEvent)} and {@link
+ * #projectFailed(ExecutionEvent)}) rather than on the end of the Maven session {@link
+ * #sessionEnded(ExecutionEvent)} because OpenTelemetry and GRPC classes are unloaded by the Maven
+ * classloader before {@link #sessionEnded(ExecutionEvent)} causing {@link NoClassDefFoundError}
+ * messages in the logs.
+ */
+@Component(role = ExecutionListener.class, hint = "otel-execution-listener")
+public class OtelExecutionListener extends AbstractExecutionListener {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  @Requirement private SpanRegistry spanRegistry;
+
+  @Requirement private OpenTelemetrySdkService openTelemetrySdkService;
+
+  /**
+   * Register in given {@link OtelExecutionListener} to the lifecycle of the given {@link
+   * MavenSession}
+   *
+   * @see org.apache.maven.execution.MavenExecutionRequest#setExecutionListener(ExecutionListener)
+   */
+  public static void registerOtelExecutionListener(
+      @Nonnull MavenSession session, @Nonnull OtelExecutionListener otelExecutionListener) {
+    @Nullable
+    ExecutionListener initialExecutionListener = session.getRequest().getExecutionListener();
+    if (initialExecutionListener instanceof ChainedExecutionListener) {
+      // already initialized
+      LoggerFactory.getLogger(OtelExecutionListener.class)
+          .debug(
+              "OpenTelemetry: OpenTelemetry extension already registered as execution listener, skip.");
+    } else {
+      session
+          .getRequest()
+          .setExecutionListener(
+              new ChainedExecutionListener(otelExecutionListener, initialExecutionListener));
+      LoggerFactory.getLogger(OtelExecutionListener.class)
+          .debug(
+              "OpenTelemetry: OpenTelemetry extension registered as execution listener. InitialExecutionListener: "
+                  + initialExecutionListener);
+    }
+  }
+
+  @Override
+  public void sessionStarted(ExecutionEvent executionEvent) {
+    MavenProject project = executionEvent.getSession().getTopLevelProject();
+    TextMapGetter<Map<String, String>> getter =
+        new TextMapGetter<Map<String, String>>() {
+          @Override
+          public Iterable<String> keys(Map<String, String> environmentVariables) {
+            return environmentVariables.keySet();
+          }
+
+          @Nullable
+          @Override
+          public String get(@Nullable Map<String, String> environmentVariables, String key) {
+            return environmentVariables == null
+                ? null
+                : environmentVariables.get(key.toUpperCase(Locale.ROOT));
+          }
+        };
+    io.opentelemetry.context.Context context =
+        W3CTraceContextPropagator.getInstance()
```

Comment:
Use `OpenTelemetry.getPropagators` instead of hard-coding the propagation format.